pkg/controller/kfservice/resources/istio/virtualservice.go

```diff
@@ -0,0 +1,260 @@
+/*
+Copyright 2019 kubeflow.org.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package istio
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha2"
+	"github.com/kubeflow/kfserving/pkg/constants"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	duckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
+	istiov1alpha1 "knative.dev/pkg/apis/istio/common/v1alpha1"
+	istiov1alpha3 "knative.dev/pkg/apis/istio/v1alpha3"
+)
+
+const (
+	IngressConfigKeyName = "ingress"
+)
+
+// Status Constants
+var (
+	PredictorSpecMissing       = "predictorSpecMissing"
+	PredictorStatusUnknown     = "predictorStatusUnknown"
+	PredictorHostnameUnknown   = "predictorHostnameUnknown"
+	TransformerStatusUnknown   = "transformerStatusUnknown"
+	TransformerHostnameUnknown = "transformerHostnameUnknown"
+	ExplainerSpecMissing       = "explainerSpecMissing"
+	ExplainerStatusUnknown     = "explainerStatusUnknown"
+	ExplainerHostnameUnknown   = "explainerHostnameUnknown"
+)
+
+type IngressConfig struct {
+	IngressGateway     string `json:"ingressGateway,omitempty"`
+	IngressServiceName string `json:"ingressService,omitempty"`
+}
+
+type VirtualServiceBuilder struct {
+	ingressConfig *IngressConfig
+}
+
+func NewVirtualServiceBuilder(config *corev1.ConfigMap) *VirtualServiceBuilder {
+	ingressConfig := &IngressConfig{}
+	if ingress, ok := config.Data[IngressConfigKeyName]; ok {
+		err := json.Unmarshal([]byte(ingress), &ingressConfig)
+		if err != nil {
+			panic(fmt.Errorf("Unable to unmarshall ingress json string due to %v ", err))
```

Comment:
nit: trailing whitespace?
"Unable to parse IngressConfig: %v"

Reply:
Updated string